Polydiv

cryptojedi opened the pull request:
constant-time polynomial division and almost constant-time polynomial inversion. Fully constant-time inversion will need some more features from TRestrictions

franziskuskiefer commented on hacspec/src/poly.rs:
```diff
@@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
-    let mut new_t = vec![T::default(); x.len()];
-    new_t[0] = T::from_literal(1);
+// Pad to degree d and reverse coefficients
+fn reverse<T: TRestrictions<T>>(x: &[T], d: usize) -> Vec<T> {
```
Maybe `#[inline]` as well?

franziskuskiefer commented on hacspec/src/poly.rs:
```diff
@@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
+fn reverse<T: TRestrictions<T>>(x: &[T], d: usize) -> Vec<T> {
+    let mut out = vec![T::default(); d+1];
+    for i in 0..std::cmp::min(x.len(), d+1) {
+        out[d-i] = x[i]; //XXX: this could probably be written more nicely with iterators?
```
You can just use `reverse` on the vector (something like `x.to_vec().reverse().truncate(d)`).
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reverse

franziskuskiefer commented on hacspec/src/poly.rs:
```diff
@@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
+    let x = monomial(T::from_literal(1),1);
+    let mut a = an.to_vec();
+    let mut b = bn.to_vec();
+    for _ in 0..cd-ad {
```
Hm, I don't like these loops but I also don't have a good idea how to make them nicer ... (maybe just ignore this comment :))

franziskuskiefer commented on hacspec/src/poly.rs:
```diff
@@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
+fn poly_divx<T: TRestrictions<T>>(v: &[T]) -> Vec<T> {
+    let mut out = vec![T::default(); v.len()-1];
+    for i in 0..out.len() {
+        out[i] = v[i+1]; //XXX: this could probably be written more nicely with iterators?
```
Something like this should work, or you can use [`drain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain)
```Rust
for (a, &b) in out.iter_mut().zip(v.iter().skip(1)) {
  *a = b;
}
```
using [skip](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip).

franziskuskiefer commented on hacspec/src/poly.rs:
```diff
@@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
+/// Iterate division steps in the constant-time polynomial inversion algorithm
+/// See Figure 5.1 from https://eprint.iacr.org/2019/266
+/// Instead of returning M2kx((u,v,q,r)) in last component, only return v
+fn divstepsx<T: TRestrictions<T>>(nn: usize, t: usize, mut delta: i128, fin: &[T], gin: &[T], n: T) -> (i128, Vec<T>, Vec<T>, (Vec<T>, usize)) {
```
Is the third argument (`delta`) really necessary? It's only called with `1`.

franziskuskiefer commented on hacspec/src/poly.rs:
```diff
@@ -61,6 +61,15 @@ fn truncate<T: TRestrictions<T>>(v: &[T]) -> Vec<T> {
 }
 
+#[inline]
+fn fixed_truncate<T: TRestrictions<T>>(v: &[T], t: usize) -> Vec<T> {
```
Oh, this also expands. The name is a little off then.

franziskuskiefer commented on hacspec/src/poly.rs:
```diff
@@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
+        g = fixed_truncate(&g, nn-i);
+
+        // TODO: make swap constant time
+        if delta > 0  && g[0] != T::default() {
```
So what do we need here to make a cswap? `greater_than_bm` and `not_equal_bm` returning all `0 `/`1` of type `T` and `BitAnd`, `BitXor`, `BitOr` should be sufficient I think.
Just making sure we have everything in `Numeric`.

cryptojedi commented:
Franziskus Kiefer <notifications@github.com> wrote:
> franziskuskiefer commented on this pull request.
> 
> Thanks! Looks pretty good. I added a couple comments. Nothing crucial.
> I also suggest using some more `Vec` functions. I'm actually not sure
> if that's a good idea or not if we want to consider a `no_std` build
> where we don't have `Vec`.

I'm using a few more Vec functions now (see below). However, I agree
that longer-term it would be nice to have a no_std build.

> Lets merge when the comments are addressed and file an issue for
> making it fully constant time. When #3 is merged we can move this to
> the new `Numeric` trait that should provide all necessary functions to
> make this fully constant time.
> 
> >      let (x, y) = normalize(x, y);
> -    let mut q = vec![T::default(); x.len()];
> -    let mut r = x.clone();
> -    let (d, c) = leading_coefficient(&y);
> -    let (mut r_d, mut r_c) = leading_coefficient(&r);
> -
> -    let mut i = 0;
> -
> -    while r_d >= d && !is_zero(&r) {
> -        let idx = r_d - d;
> -
> -        let c_idx = if n == T::default() {
> +    let mut rem = x.clone();
> +    let mut quo = vec![T::default(); x.len()];
> +    let (yd, c) = leading_coefficient(&y);
> +    let dist = x.len() - yd; // length of x and degree of y are assumed to be public
> 
> Can you add this assumption to the function comment above? Then it gets put into the documentation.

Done.

> >              // In ℤ we try this. It might not work.
> -            r_c / c
> +            rem[idx] / c
> 
> So this won't work with `Numeric` in #3, which only implements
> operators that are "safe" to use, i.e. no `/` (div). This would need
> to be `rem[idx].div(c)` then, where `div` is a constant-time
> implementation for secret `Numeric`s. Can you add a comment here so
> that's on the radar when moving to `Numeric`?

Comment added.

> >              // In ℤ we try this. It might not work.
> -            r_c / c
> +            rem[idx] / c
>          } else {
>              // r_c / c in ℤn is r_c * 1/c.
> 
> The comment needs updating

Fixed.

> >          };
> -        if c_idx == T::default() {
> -            panic!("c_idx is 0; can't divide these two polynomials");
> +        if t == T::default() && rem[idx] != T::default() {
> 
> Same as the `div` comment.

Are you referring to the panic message? That's updated now.

> > @@ -61,6 +61,15 @@ fn truncate<T: TRestrictions<T>>(v: &[T]) -> Vec<T> {
>      out
>  }
>  
> +#[inline]
> +fn fixed_truncate<T: TRestrictions<T>>(v: &[T], t: usize) -> Vec<T> {
> 
> I think the other `truncate` is unused at this point. Can you remove it?

Done.

> > @@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
>      t
>  }
>  
> -/// Extended euclidean algorithm to compute the inverse of x in yℤ[x]
> -#[inline]
> -pub fn extended_euclid<T: TRestrictions<T>>(x: &[T], y: &[T], n: T) -> Result<Vec<T>, &'static str> {
> -    let (x, y) = normalize(x, y);
> -    println!("euclid: {:?} / {:?}", x, y);
> -
> -    let mut new_t = vec![T::default(); x.len()];
> -    new_t[0] = T::from_literal(1);
> +// Pad to degree d and reverse coefficients
> +fn reverse<T: TRestrictions<T>>(x: &[T], d: usize) -> Vec<T> {
> 
> Maybe `#[inline]` as well?

It's removed now; using the reverse() function of Vec instead.

> > @@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
>      t
>  }
>  
> -/// Extended euclidean algorithm to compute the inverse of x in yℤ[x]
> -#[inline]
> -pub fn extended_euclid<T: TRestrictions<T>>(x: &[T], y: &[T], n: T) -> Result<Vec<T>, &'static str> {
> -    let (x, y) = normalize(x, y);
> -    println!("euclid: {:?} / {:?}", x, y);
> -
> -    let mut new_t = vec![T::default(); x.len()];
> -    new_t[0] = T::from_literal(1);
> +// Pad to degree d and reverse coefficients
> +fn reverse<T: TRestrictions<T>>(x: &[T], d: usize) -> Vec<T> {
> +    let mut out = vec![T::default(); d+1];
> +    for i in 0..std::cmp::min(x.len(), d+1) {
> +        out[d-i] = x[i]; //XXX: this could probably be written more nicely with iterators?
> 
> You can just use `reverse` on the vector (something like `x.to_vec().reverse().truncate(d)`).
> https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reverse
 
Need to truncate/pad first (see below) and then reverse. Done.

> -    let mut new_r = x.clone();
> +/// Subtract quotient (bn/x^bd) from (an/x^ad)
> +fn quot_sub<T: TRestrictions<T>>(an: &[T], ad: usize, bn: &[T], bd: usize, n: T) -> (Vec<T>, usize){
> +    let cd = std::cmp::max(ad, bd);
> +    let x = monomial(T::from_literal(1),1);
> +    let mut a = an.to_vec();
> +    let mut b = bn.to_vec();
> +    for _ in 0..cd-ad {
> 
> Hm, I don't like these loops but I also don't have a good idea how to
> make them nicer ... (maybe just ignore this comment :))

I added a comment that this might want beautification.

> -    let mut t = vec![T::default(); x.len()];
> -    let mut r = y.clone();
> +/// Divide a by x assuming a is a multiple of x (shift right by one)
> +fn poly_divx<T: TRestrictions<T>>(v: &[T]) -> Vec<T> {
> +    let mut out = vec![T::default(); v.len()-1];
> +    for i in 0..out.len() {
> +        out[i] = v[i+1]; //XXX: this could probably be written more nicely with iterators?
> 
> Something like this should work, or you can use [`drain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain)
> ```Rust
> for (a, &b) in out.iter_mut().zip(v.iter().skip(1)) {
>   *a = b;
> }
> ```
> using [skip](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip).

Nice! Done.

> >  
> -    while !is_zero(&new_r) {
> -        println!("{:?} / {:?}", r, new_r);
> -        let q = euclid_div(&r, &new_r, n).0;
> +/// Iterate division steps in the constant-time polynomial inversion algorithm
> +/// See Figure 5.1 from https://eprint.iacr.org/2019/266
> +/// Instead of returning M2kx((u,v,q,r)) in last component, only return v
> +fn divstepsx<T: TRestrictions<T>>(nn: usize, t: usize, mut delta: i128, fin: &[T], gin: &[T], n: T) -> (i128, Vec<T>, Vec<T>, (Vec<T>, usize)) {
> 
> Is the third argument (`delta`) really necessary? It's only called with `1`.

Good point. It's an artifact from the paper where the algorithm comes
from and where it's used also in other contexts. I removed the argument
delta now.

> > @@ -61,6 +61,15 @@ fn truncate<T: TRestrictions<T>>(v: &[T]) -> Vec<T> {
>      out
>  }
>  
> +#[inline]
> +fn fixed_truncate<T: TRestrictions<T>>(v: &[T], t: usize) -> Vec<T> {
> 
> Oh, this also expands. The name is a little off then.

Indeed. Renamed to "make_fixed_length" now. Could also use something
like pad_truncate? Whatever you prefer.

> > +    let mut q = (vec![T::default(); 1], 0);
> +    let mut r = (vec![T::from_literal(1); 1], 0);
> +
> +    for i in 0..nn {
> +        // Bring u,v,q,r back to fixed precision t
> +        u.0 = fixed_truncate(&u.0, t);
> +        v.0 = fixed_truncate(&v.0, t);
> +        q.0 = fixed_truncate(&q.0, t);
> +        r.0 = fixed_truncate(&r.0, t);
> +
> +        // Decrease precision of f and g in each iteration
> +        f = fixed_truncate(&f, nn-i);
> +        g = fixed_truncate(&g, nn-i);
> +
> +        // TODO: make swap constant time
> +        if delta > 0  && g[0] != T::default() {
> 
> So what do we need here to make a cswap? `greater_than_bm` and
> `not_equal_bm` returning all `0 `/`1` of type `T` and `BitAnd`,
> `BitXor`, `BitOr` should be sufficient I think.

Sounds good.

> Just making sure we have everything in `Numeric`.

Not sure how generic Numeric is going to be. We additionally need a way
to turn the boolean condition (delta>0) into T. We cannot work with a
boolean there, but have to work with an int (that's either 0 or 1).
